mne/preprocessing/xdawn.py

```diff
@@ -1,266 +1,362 @@
+        epochs overlap.
 
-    n_samples = stop - start
-    n_epochs, n_channels, n_times = epochs_data.shape
-    events_pos = events[:, 0] - events[0, 0]
 
-    data = np.zeros((n_channels, n_samples))
-    for idx in range(n_epochs):
-        onset = events_pos[idx]
-        offset = onset + n_times
-        data[:, onset:offset] = epochs_data[idx]
+    Attributes
+    ----------
+    filters_ : array, shape (n_channels, n_channels)
+        The Xdawn components used to decompose the data for each event type.
+    patterns_ : array, shape (n_channels, n_channels)
+        The Xdawn patterns used to restore the signals for each event type.
 
-    return data
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
 
 
-def _least_square_evoked_epochs(epochs_data, events, event_id, tmin,
-                                tmax, sfreq, return_toeplitz=False):
-    """Least square estimation of evoked response from a Epochs instance.
+    See Also
+    --------
+    CSP
+    XDawn
+    """
+    n_epochs, n_channels, n_times = epochs_data.shape
 
-    Parameters
-    ----------
-    epochs : Epochs instance
-        An instance of Epochs.
-    return_toeplitz : bool (default False)
-        If true, compute the toeplitz matrix.
+    classes = np.unique(y)
+
+    # Retrieve or compute whitening covariance
+    if signal_cov is None:
+        signal_cov = _regularized_covariance(np.hstack(epochs_data), reg)
+    elif isinstance(signal_cov, Covariance):
+        signal_cov = signal_cov.data
+    if not isinstance(signal_cov, np.ndarray) or (
+            not np.array_equal(signal_cov.shape,
+                               np.tile(epochs_data.shape[1], 2))):
+        raise ValueError('signal_cov must be None, a covariance instance '
+                         'or a array of shape (n_chans, n_chans)')
+
+    # Get prototype events
+    if events is not None:
+        evokeds, toeplitzs = _least_square_evoked(
+            epochs_data, events, tmin, sfreq)
+    else:
+        evokeds, toeplitzs = list(), list()
+        for c in classes:
+            # Prototyped response for each class
+            evokeds.append(np.mean(epochs_data[y == c, :, :], axis=0))
+            toeplitzs.append(1.)
+
+    filters = list()
+    patterns = list()
+    for evo, toeplitz in zip(evokeds, toeplitzs):
+        # Covariance matrix of the prototyper response & signal
+        evo = np.dot(evo, toeplitz)
+        evo_cov = np.matrix(_regularized_covariance(evo, reg))
+
+        # Spatial filters
```

Comment:
Can you clarify if this *constructs* spatial filters or *applies* spatial filters? (I think the former right?)